### ISSUE TITLE
Expose DebugConfigurationProvider for testing

### ISSUE THREAD

**Issue by DanTup** [Microsoft/vscode#46028]
I'd like to write a test that allows me to simulate a user pressing `F5` (eg. calls the DebugConfigurationProvider to resolve a debug config) and then drive the debug session using `vscode-debugadapter-testsupport`.

I've checked all of the commands returned from `getCommands` and the APIs on `vscode.debug` but can't see anything suitable.

I think the bit missing, is being able to call `resolveDebugConfig`. If there was a command for this, I think I would be able to do all of this:

```ts
const dc = new DebugClient("node", "./out/src/debug/dart_debug_entry.js", "dart");
const debugConfig = vs.commands.executeCommand("#??????????#", {
  type: "dart",
  request" launch",
  name: "Dart"
});
dc.start(debugConfig.debugServer);
await Promise.all([
	dc.configurationSequence(),
	dc.launch(debugConfig),
	dc.waitForEvent("terminated"),
]);
```

The reason I need this is that my `resolveDebugConfiguration` method attaches a bunch of things to the config that are required for the debugger to work (such as SDK paths). I don't have any easy way to get all of these from inside my test.

Note: The above code assumes that the resolve method is setting `debugServer` to run in-process (which is what my plan is based on prior discussions), but presumably if that's not the case then `debugServer` could just be added to the original base config passed into this new command.

(cc @weinand)

**Comment by DanTup** [comment]
Er, it's posisble `vs.debug.startDebugging` is already doing this - I'm working on a branch that doesn't have my `debugServer` changes which may be why it isn't coming through... I'll fix this up and test and post back!

**Comment by DanTup** [comment]
@weinand Seems like this almost worked... I did this:

```ts
it("runs a Dart script to completion", async () => {
	await vs.debug.startDebugging(vs.workspace.workspaceFolders[0], debugConfig);
	console.log(debugConfig.debugServer); // This is undefined :(
	await dc.start(debugConfig.debugServer);
	await dc.waitForEvent("terminated");
});
```

However, despite my `resolveDebugConfiguration` mutating the config it's given, `debugConfig` doesn't have any of the additional properties that were added to it after the `startDebugging` call. Any ideas?

**Comment by weinand** [comment]
@DanTup yes, the debugConfig passed to `vs.debug.startDebugging` is input only. If you add attributes in resolveDebugConfiguration they do not show up after `vs.debug.startDebugging` returns.

**Comment by DanTup** [comment]
@weinand Is there any way around this so I can get at the return from `resolve` inside my test? (short of waiting for something to be implemented in future?)

**Comment by DanTup** [comment]
@weinand I was able to get around the original issue here by just assigning a port number in my test before calling `vs.debug.startDebugging` like this:

```ts
describe.only("debugger", () => {
	const dc = new DebugClient("node", "./out/src/debug/dart_debug_entry.js", "dart");
	const debugConfig: vs.DebugConfiguration = {
		debugServer: 4711,
		name: "Dart & Flutter",
		program: "${workspaceFolder}/bin/main.dart",
		request: "launch",
		type: "dart",
	};

	before(() => activate());

	it("runs a Dart script to completion", async () => {
		await vs.debug.startDebugging(vs.workspace.workspaceFolders[0], debugConfig);
		await dc.start(debugConfig.debugServer);
		console.log("Waiting for terminated");
		await dc.waitForEvent("terminated");
		console.log("Done!");

	});
});
```

The problem is, this hands on the `waitForEvent` line. It doesn't trigger the internal timeout (I used to see "timed out weighting for terminated") but rather just sits there until the test times out.

I'm now wondering whether it's valid for me to have two debug clients (the one inside Code that's starting when I call `vs.debug.startDebugging` and the test one I connect on the next line). I did see that [Chrome supports it](https://code.visualstudio.com/blogs/2017/12/20/chrome-debugging) so I wonder whether there is some way I could make this work?

**Comment by DanTup** [comment]
@weinand The title of this case may be a little misleading now, as it may just be a question about whether what I'm doing is valid or not (and if so, why it doesn't work).

**Comment by DanTup** [comment]
(If it's not valid, then the original request stands - I've ended up exporting my debug config provider in the activate method for now, which seems to be doing the job of at least letting me run my DA :-))

**Comment by connor4312** [comment]
It sounds like you figured out how to make this work in the intervening 6 years :) I'm going to go ahead and close this issue for housekeeping, but please let us know if you have followup questions.
